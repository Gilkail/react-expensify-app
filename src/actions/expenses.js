import uuid from 'uuid'
import database from '../firebase/firebase'


// Add expense function which will trigger the addExpense action called from AddExpansePage
export const startAddExpense = (expenseData = {}) => { // Object transfered from 
    return (dispatch) => { // Transfering action dispatch due to thunk configured in store config file
        const {description = '', note = '', amount = 0, createdAt = 0} = expenseData // Destructuring object and setting up defaults
        const expense = {description, note, amount, createdAt} // Passing variables to expense array
        return database.ref('expenses').push(expense) // Pushing the data to firebase
            .then((ref)=>{ // Getting promise from fetched data with the ref
                dispatch(addExpense({ // Calling the addExpense dispatch
                    id: ref.key, // Taking the generated key from the fetched data using ref.key
                    ...expense // Spreading the expense object and passing to the action generator
                }))
        })
    }
}

// Add expnese
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

// Remove expense
export const removeExpense = ({id} = {}) =>({
    type: 'REMOVE_EXPENSE',
    id
})


// Edit expense
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
})

// export const startSetExpenses

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot)=>{
            const expenses = []
            snapshot.forEach((expense)=>{
                expenses.push({
                    id: expense.key,
                    ...expense.val()
                })
            })
            dispatch(setExpenses(expenses))
        })
    }
}