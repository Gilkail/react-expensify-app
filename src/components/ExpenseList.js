import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div>
        <h2>Expense List</h2>
        {
            props.expenses.length === 0 ? (
                <p>no expenses</p>
            ) : (
                props.expenses.map((expense, index)=>(
                    <div key={index}>
                    <ExpenseListItem {...expense}/>
                    </div>
                ))
            )
        }
    </div>
)

const mapStateToProps = (state)=>{
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList)