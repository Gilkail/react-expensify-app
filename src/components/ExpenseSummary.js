import React from 'react'
import numeral from 'numeral'
import {connect} from 'react-redux'
import selectExpenses from '../selectors/expenses'
import expensesTotal from '../selectors/expenses-total'

export const ExpenseSummary = (props) => (
    <div>
    {
        <h2> Viewing {props.expenses.length} {props.expenses.length === 1? 'expense' : 'expenses'} totalling {numeral(expensesTotal(props.expenses)/100).format('$0,0.00')}</h2>
    }      
    </div>
)

const mapStateToProps = (state)=>{
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseSummary)