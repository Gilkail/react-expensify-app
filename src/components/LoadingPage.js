import React from 'react'

const LoadingPage = () => (
    <div className="loader">
        <img class="loader__image" src="/img/loader.gif" />
    </div>
)

export default LoadingPage