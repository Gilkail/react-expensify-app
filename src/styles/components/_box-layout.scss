.box-layout{
    background: url('/img/bg.jpg');
    background-size: cover;
    background-position: center;
    height: 100vh;
    width: 100vw;
    display: flex;
    justify-content: center;
    align-items: center;
}

.box-layout__box{
    background: fade-out(white, .15);
    border-radius: 3px;
    text-align: center;
    padding: $l-size $m-size;
    width: 40rem;
}

.box-layout__title{
    margin: 0 0 $s-size;
    line-height: 1;
}