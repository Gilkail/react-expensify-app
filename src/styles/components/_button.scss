.button{
    padding: $s-size $m-size;
    font-size: $font-size-large;
    background-color: $cta-blue;
    border: none;
    display: inline-block;
    color: white;
    line-height: 1;
    font-weight: 300;
    text-decoration: none;
}

.button--secondary{
    background-color: lighten($dark-gray, 20%);
}

.button--link{
    background: none;
    border: none;
}