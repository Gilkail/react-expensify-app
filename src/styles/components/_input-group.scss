.input-group{
    display: flex;
    margin-bottom: $m-size;
    flex-direction: column;

    @media (min-width: $mobile-break-point){
        flex-direction: row;
        margin-bottom: $l-size;
    }
}

.input-group__item{
    margin-right: $s-size;
    margin-bottom: $s-size;
    @media (min-width: $mobile-break-point){
        margin: 0 $s-size 0 0;
    }
}