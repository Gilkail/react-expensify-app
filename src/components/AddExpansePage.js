import React from 'react'
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import {addExpense} from '../actions/expenses'

export class AddExpansePage extends React.Component{
    onSubmit = (expense) =>{
        this.props.addExpense(expense)
        this.props.history.push('/')
    }
    render(){
        return (
            <div>
                <h1>This is my add expense component</h1>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        )
    }
}

// This function allows to to use dispatch in more simplified way inside the componenet in order to test it
const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
})

// Inside the connect function the "undefined" is for the mapStateToProps in order the mapDispatchToProps will be usable
export default connect( undefined , mapDispatchToProps)(AddExpansePage)