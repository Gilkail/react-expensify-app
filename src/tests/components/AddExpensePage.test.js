import React from 'react'
import {shallow} from 'enzyme'
import {AddExpansePage} from '../../components/AddExpansePage'
import expenses from '../fixtures/expenses'

let addExpenseSpy, history, wrapper

beforeEach(()=>{
    addExpenseSpy = jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(<AddExpansePage addExpense={addExpenseSpy} history={history}/>)
})

test('should render addExpensePage correctly', ()=>{
    expect(wrapper).toMatchSnapshot()
})

test('should handle onsubmit', ()=>{
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0])
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(addExpenseSpy).toHaveBeenLastCalledWith(expenses[0])
})