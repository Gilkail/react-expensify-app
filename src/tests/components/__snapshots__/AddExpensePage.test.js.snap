// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render addExpensePage correctly 1`] = `
<div>
  <h1>
    This is my add expense component
  </h1>
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
